src/dmd/target.d

```diff
@@ -887,5 +897,42 @@ struct TargetObjC
     }
 }
 
+/// Support for `__DATE__`, `__TIME__`, and `__TIMESTAMP__`
+struct TimeStampInfo
+{
+    private __gshared bool initdone = false;
+
+    // Note: Those properties need to be guarded by a call to `init`
+    // The API isn't safe, and quite brittle, but it was left this way
+    // over performance concerns.
+    // This is currently only called once, from the lexer.
+    __gshared char[11 + 1] date;
+    __gshared char[8 + 1] time;
+    __gshared char[24 + 1] timestamp;
+
+    public void init(const ref Loc loc) nothrow
+    {
+        if (initdone)
+            return;
+
+        initdone = true;
+        time_t ct;
+        // https://issues.dlang.org/show_bug.cgi?id=20444
+        if (auto p = getenv("SOURCE_DATE_EPOCH"))
+        {
+            if (!ct.parseDigits(p.toDString()))
+                error(loc, "Value of environment variable `SOURCE_DATE_EPOCH` should be a valid timestamp, not: `%s`", p);
```

Comment:
for usability I think this should say "valid unix timestamp" or "valid integer timestamp"